File: ThreeDSecure/src/main/java/com/braintreepayments/api/CardinalClient.java
package com.braintreepayments.api;

import android.content.Context;

import androidx.fragment.app.FragmentActivity;

import com.cardinalcommerce.cardinalmobilesdk.Cardinal;
import com.cardinalcommerce.cardinalmobilesdk.enums.CardinalEnvironment;
import com.cardinalcommerce.cardinalmobilesdk.enums.CardinalRenderType;
import com.cardinalcommerce.cardinalmobilesdk.enums.CardinalUiType;
import com.cardinalcommerce.cardinalmobilesdk.models.CardinalChallengeObserver;
import com.cardinalcommerce.cardinalmobilesdk.models.CardinalConfigurationParameters;
import com.cardinalcommerce.cardinalmobilesdk.models.ValidateResponse;
import com.cardinalcommerce.cardinalmobilesdk.services.CardinalInitService;
import com.cardinalcommerce.cardinalmobilesdk.services.CardinalValidateReceiver;

import org.json.JSONArray;

class CardinalClient {

    private String consumerSessionId;

    CardinalClient () {}

    void initialize(Context context, Configuration configuration, final ThreeDSecureRequest request, final CardinalInitializeCallback callback) throws BraintreeException {
        configureCardinal(context, configuration, request);

        CardinalInitService cardinalInitService = new CardinalInitService() {
            @Override
            public void onSetupCompleted(String sessionId) {
                consumerSessionId = sessionId;
                callback.onResult(consumerSessionId, null);
            }

            @Override
            public void onValidated(ValidateResponse validateResponse, String serverJWT) {
                if (consumerSessionId == null) {
                    callback.onResult(null, new BraintreeException("consumer session id not available"));
                } else {
                    callback.onResult(consumerSessionId, null);
                }
            }
        };

        try {
            Cardinal.getInstance().init(configuration.getCardinalAuthenticationJwt(), cardinalInitService);
        } catch (RuntimeException e) {
            throw new BraintreeException("Cardinal SDK init Error.", e);
        }
    }

    void continueLookup(ThreeDSecureResult threeDSecureResult, CardinalChallengeObserver challengeObserver) throws BraintreeException {
        ThreeDSecureLookup lookup = threeDSecureResult.getLookup();
        String transactionId = lookup.getTransactionId();
        String paReq = lookup.getPareq();
        try {
            Cardinal.getInstance().cca_continue(transactionId, paReq, challengeObserver);
        } catch (RuntimeException e) {
            throw new BraintreeException("Cardinal SDK cca_continue Error.", e);
        }
    }

    private void configureCardinal(Context context, Configuration configuration, ThreeDSecureRequest request) throws BraintreeException {
        CardinalEnvironment cardinalEnvironment = CardinalEnvironment.STAGING;
        if ("production".equalsIgnoreCase(configuration.getEnvironment())) {
            cardinalEnvironment = CardinalEnvironment.PRODUCTION;
        }

        CardinalConfigurationParameters cardinalConfigurationParameters = new CardinalConfigurationParameters();
        cardinalConfigurationParameters.setEnvironment(cardinalEnvironment);
        cardinalConfigurationParameters.setRequestTimeout(8000);
        cardinalConfigurationParameters.setEnableDFSync(true);

        switch (request.getUiType()) {
            case ThreeDSecureRequest.NATIVE:
                cardinalConfigurationParameters.setUiType(CardinalUiType.NATIVE);
            case ThreeDSecureRequest.HTML:
                cardinalConfigurationParameters.setUiType(CardinalUiType.HTML);
            case ThreeDSecureRequest.BOTH:
                cardinalConfigurationParameters.setUiType(CardinalUiType.BOTH);
        }

        if (request.getRenderTypes() != null) {
            JSONArray renderTypes = new JSONArray();

            for (Integer renderType : request.getRenderTypes()) {
                if (renderType.equals(ThreeDSecureRequest.OTP)) {
                    renderTypes.put(CardinalRenderType.OTP);
                } else if (renderType.equals(ThreeDSecureRequest.SINGLE_SELECT)) {
                    renderTypes.put(CardinalRenderType.SINGLE_SELECT);
                } else if (renderType.equals(ThreeDSecureRequest.MULTI_SELECT)) {
                    renderTypes.put(CardinalRenderType.MULTI_SELECT);
                } else if (renderType.equals(ThreeDSecureRequest.OOB)) {
                    renderTypes.put(CardinalRenderType.OOB);
                } else if (renderType.equals(ThreeDSecureRequest.RENDER_HTML)) {
                    renderTypes.put(CardinalRenderType.HTML);
                }
            }

            cardinalConfigurationParameters.setRenderType(renderTypes);
        }

        if (request.getV2UiCustomization() != null) {
            cardinalConfigurationParameters.setUICustomization(request.getV2UiCustomization().getCardinalUiCustomization());
        }

        try {
            Cardinal.getInstance().configure(context, cardinalConfigurationParameters);
        } catch (RuntimeException e) {
            throw new BraintreeException("Cardinal SDK configure Error.", e);
        }
    }

    String getConsumerSessionId() {
        return consumerSessionId;
    }

    void cleanup() {
        Cardinal.getInstance().cleanup();
    }
}

File: ThreeDSecure/src/main/java/com/braintreepayments/api/CardinalClient.kt
package com.braintreepayments.api

import android.content.Context
import com.cardinalcommerce.ThreeDotOh.CardinalService
import com.cardinalcommerce.ThreeDotOh.models.CardinalChallengeObserver
import com.cardinalcommerce.ThreeDotOh.models.CardinalChallengeParameters
import com.cardinalcommerce.ThreeDotOh.models.CardinalConfigurationParameters
import com.cardinalcommerce.ThreeDotOh.models.enums.CardinalEnvironment
import com.cardinalcommerce.ThreeDotOh.models.enums.CardinalRenderType
import com.cardinalcommerce.ThreeDotOh.models.enums.CardinalUiType
import com.cardinalcommerce.shared.cs.interfaces.Error
import org.json.JSONArray

internal class CardinalClient {
    var consumerSessionId: String? = null
        private set

    @Throws(BraintreeException::class)
    fun initialize(
        context: Context,
        configuration: Configuration,
        request: ThreeDSecureRequest,
        callback: CardinalInitializeCallback
    ) {
        // TODO: uncomment once PRODUCTION enum is visible to SDK
//        val environment = if ("production".equals(configuration.environment, ignoreCase = true)) {
//            // NOTE: PRODUCTION enum not visible to SDK
//            CardinalEnvironment.PRODUCTION
//        } else {
//            CardinalEnvironment.STAGING
//        }
        val cardinalEnvironment = CardinalEnvironment.STAGING
        val cardinalUIType = when (request.uiType) {
            ThreeDSecureRequest.NATIVE -> CardinalUiType.NATIVE
            ThreeDSecureRequest.HTML -> CardinalUiType.HTML
            else -> CardinalUiType.BOTH
        }

        val cardinalRenderTypes = JSONArray()
        request.renderTypes?.let { renderTypes ->
            for (renderType in renderTypes) {
                val cardinalRenderType = when (renderType) {
                    ThreeDSecureRequest.OTP -> CardinalRenderType.OTP
                    ThreeDSecureRequest.SINGLE_SELECT -> CardinalRenderType.SINGLE_SELECT
                    ThreeDSecureRequest.MULTI_SELECT -> CardinalRenderType.MULTI_SELECT
                    ThreeDSecureRequest.OOB -> CardinalRenderType.OOB
                    ThreeDSecureRequest.RENDER_HTML -> CardinalRenderType.HTML
                    else -> null
                }
                cardinalRenderType?.let { cardinalRenderTypes.put(it) }
            }
        }

        val cardinalParams = CardinalConfigurationParameters().apply {
            environment = cardinalEnvironment
            // QUESTION: is this supposed to be requestTimeout?
            sdkMaxTimeout = 8000
            // QUESTION: is dfsync property no longer needed?
            uiType = cardinalUIType
            renderType = cardinalRenderTypes
        }

        request.v2UiCustomization?.cardinalUiCustomization?.let {
            cardinalParams.uiCustomization = it
        }

        val cardinalService = CardinalService.getInstance()
        val cardinalCallback =
            object : com.cardinalcommerce.ThreeDotOh.interfaces.CardinalInitializeCallback {
                override fun onSuccess(sdkTransactionId: String?) {
                    // QUESTION: is SDK transaction ID the same as Consumer Session ID?
                    consumerSessionId = sdkTransactionId
                    callback.onResult(consumerSessionId, null)
                }

                override fun onError(error: Error<Int>?) {
                    val braintreeException = error?.let {
                        BraintreeException("An 3DS initialize error occurred. Cardinal Error code: $it")
                    }
                    braintreeException?.let { btError ->
                        callback.onResult(null, btError)
                    }
                }
            }
        val serverJWT = configuration.cardinalAuthenticationJwt
        cardinalService.initialize(context, serverJWT, cardinalParams, cardinalCallback)
    }

    @Throws(BraintreeException::class)
    fun continueLookup(
        threeDSecureResult: ThreeDSecureResult,
        challengeObserver: CardinalChallengeObserver?
    ) {
        val lookup = threeDSecureResult.lookup
        val transactionId = lookup.transactionId
        val paReq = lookup.pareq

        val cardinalService = CardinalService.getInstance()
        try {
            // QUESTION: Original integration required two parameters, what are these new required values?
            val challengeParams = CardinalChallengeParameters(
                "",
                "",
                "",
                "",
                "",
                ""
            )
            // QUESTION: What does the timeout parameter do? Is there a good default value we should use?
            cardinalService.doChallenge(challengeObserver, challengeParams, 80000)
        } catch (e: RuntimeException) {
            throw BraintreeException("Cardinal SDK cca_continue Error.", e)
        }
        cardinalService.cleanup()
    }
}